lib/linked_list.rb

```diff
@@ -12,120 +12,265 @@ def initialize(value, next_node = nil)
+  # returns true if found, false otherwise
+  def search(value)
+    current = @head
+    
+    until current.nil?
+      return true if current.data == value
+      current = current.next
     end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    if @head == nil
+      return nil
     end
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    def search(value)
-      raise NotImplementedError
+    
+    current = @head
+    max = @head.data
+    
+    until current == nil
+      if current.data > max
+        max = current.data
+      end
+      current = current.next
     end
-
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    def find_max
-      raise NotImplementedError
+    return max
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    if @head == nil
+      return nil
     end
-
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
+    
+    current = @head
+    min = @head.data
+    
+    until current == nil
+      if current.data < min
+        min = current.data
+      end
+      current = current.next
     end
-
-
-    # method that returns the length of the singly linked list
-    def length
-      raise NotImplementedError
+    return min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    if @head.nil?
+      return 0
     end
-
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    def get_at_index(index)
-      raise NotImplementedError
+    
+    current = @head
+    count = 0
+    
+    until current.nil?
+      count += 1
+      current = current.next
     end
-
-    # method to print all the values in the linked list
-    def visit
-      raise NotImplementedError
+    return count
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    current = @head
+    
+    index.times do
+      return nil if current.nil?
+      current = current.next
     end
-
-    # method to delete the first node found with specified value
-    def delete(value)
-      raise NotImplementedError
+    
+    return current.data

```

Comment:
current might be nil

```suggestion
    return nil if current.nil?
    return current.data
```